content/docs/reference/how.md

```diff
@@ -7,44 +7,89 @@ menu:
+The following diagram illustrates the interaction between these parts of the system:
 
 <img src="/images/docs/reference/engine-block-diagram.png" alt="Pulumi engine and providers" width="600">
 
-For instance, suppose we have the following Pulumi program, which creates two S3 buckets:
+Let's talk a little bit about each of these components and then see how they all fit together during an invocation of `pulumi up`
+
+## Language Hosts
+
+The _language host_ is responsible for running a Pulumi program and setting up an environment where it can register resources with the _deployment engine_.  The language host is made up of two different pieces:
+
+1. A language executor, which is a binary named `pulumi-language-<language-name>`, which Pulumi uses to the language your program is written in (e.g. Node or Python). This binary is distributed with the Pulumi CLI.
+2. A language runtime, which is responsible for preparing your program to be executed and observes its execution in order to detect resource registrations.  When a resource is _registered_ (via `new Resource()` in JavaScript or `Resource(...)` in Python) the language runtime communicates this request back to the _deployment engine_. The language runtime is distributed as a regular package, just like any other code you might depend on it your program.  For example, the Node runtime is contained in the `@pulumi/pulumi` package published to NPM and the Python runtime is contained in the `pulumi` package published to PyPI.
+
+## Deployment Engine
+
+The _deployment engine_ is responsible for computing the set of operations needed to drive the current state of your infrastructure into the desired state expressed by your program.  When a _resource registration_ is received from the language host, the engine consults the existing state to determine if that resource has been created before. If it has not, the engine uses a _resource provider_ in order to create it.  If it already exists, the engine works with the resource provider to determine what, if anything, has changed by comparing the old state of the resource with the new desired state of the resource as expressed by the program. If there are changes, the engine determines if it can _update_ the resource in place or if it must _replace_ it by _creating_ a new version and _deleting_ the old version. The decision depends on what properties of the resource are changing and the type of the resource itself.  When the language host communicates to the engine that it has completed execution of the Pulumi program, the engine looks for any existing resources that exist but it did not see a new resource registration for and schedules these resources for deletion.
```

Comment:
```suggestion
The _deployment engine_ is responsible for computing the set of operations needed to drive the current state of your infrastructure into the desired state expressed by your program.  When a _resource registration_ is received from the language host, the engine consults the existing state to determine if that resource has been created before. If it has not, the engine uses a _resource provider_ in order to create it.  If it already exists, the engine works with the resource provider to determine what, if anything, has changed by comparing the old state of the resource with the new desired state of the resource as expressed by the program. If there are changes, the engine determines if it can _update_ the resource in place or if it must _replace_ it by _creating_ a new version and _deleting_ the old version. The decision depends on what properties of the resource are changing and the type of the resource itself.  When the language host communicates to the engine that it has completed execution of the Pulumi program, the engine looks for any existing resources that it did not see a new resource registration for and schedules these resources for deletion.
```